snippets-directory.kak

```diff
@@ -9,82 +9,101 @@ define-command snippets-directory-disable %{
-                    firstline=0
-                    while read -r line; do
-                        if [ "$firstline" -eq 0 ]; then
-                            firstline=1
-                        else
-                            printf "\n"
-                        fi
-                        doubleupsinglequotes "$line" 4
-                    done < "$snippet"
-                    printf "'''' ''"
-                done
-                [ $first -eq 1 ] && printf "'\n"
-                cd ..
-            done
-        )
-        done
+        cat <<'EOF' | perl - "$@"
+use strict;
+use warnings;
+
+foreach my $dir (@ARGV) {
+    if(! -d $dir) {
+        print "echo -debug 'Snippets directory ''$dir'' does not exist'\n";
+        next;
+    }
+
+    opendir(my $dh, $dir) || die "Can't open $dir: $!";
+    my @filetypes = readdir($dh);
+    closedir($dh);
+    foreach my $filetype (@filetypes) {
+        next if ($filetype eq '.' || $filetype eq '..');
+        $filetype = "$dir/$filetype";
+
+        next if (! -d $filetype);
+        print_kak_commands_for_filetype_dir($filetype);
+    }
+}
+
+sub print_kak_commands_for_filetype_dir {
+    my $filetype = shift;
+    my $printed_filetype_intro = 0;
+    opendir(my $dh, $filetype) || die "Can't open $filetype: $!";
+    my @snippets = readdir($dh);
+    closedir($dh);
+    foreach my $snippet (@snippets) {
+        next if ($snippet eq '.' || $snippet eq '..');
+        my $name = $snippet;
+        my $trigger = $snippet;
+        $snippet = "$filetype/$snippet";
+
+        next if (! -f $snippet);
+        $name =~ s/.*? - //;
+        next if ($name eq "");               # no valid snippet name
+        if("$filetype/$name" eq $snippet) {  # no ' - ' in filename -> no trigger
+            $trigger = "";
+        } else {
+            $trigger =~ s/ - .*//;
+        }
+
+        if($printed_filetype_intro == 0) {
+            print_filetype_intro($filetype);
+            $printed_filetype_intro = 1;
+        }
+        print_snippet($trigger, $name, $snippet);
+    }
+    print_filetype_outro() if ($printed_filetype_intro == 1);
+}
+
+sub print_filetype_intro {
+    my $filetype = shift;
+    print "hook -group 'snippets-directory' global BufSetOption 'filetype=";
+    $filetype =~ s/.*\///;  # Poor mans basename() to avoid the extra import
+    print multiply_single_quotes($filetype, 1);
+    print "' 'set -add buffer snippets"
+}
+
+sub print_snippet {
+    my $trigger = shift;
+    my $name = shift;
+    my $snippet = shift;
+    print " ''", multiply_single_quotes($name, 2), "''";
+    print " ''", multiply_single_quotes($trigger, 2), "''";
+    print " ''snippets-insert ''''";
+    my $first_line = 1;
+    open(my $fh, "<", $snippet) || die "Can't open < $snippet: $!";
+    while (<$fh>) {
+        print "\n" if (! $first_line);
+        chomp;
+        print multiply_single_quotes($_, 4);
+        $first_line = 0;
```

Comment:
I think this part is not necessary anymore, I had it this way because of shell limitations (`read` doesn't give you the trailing newline) but here you should be able to get rid of the `first_line` stuff.

Reply:
You are right, I just blindly transcribed the code. The last newline still needs to be removed, though.